### PR TITLE
Support large files

### PR DESCRIPTION
**Pull request by Ceredron**
## Description
Added support of very large files when virus scan is disabled.

## Related Issue(s)
- #525 

## Verification
- [X] **Your** code builds clean without any errors or warnings
- [X] Manual testing done (required)
- [X] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [X] All tests run green

## Documentation
- [X] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

## Release Notes

- **New Features**
  - Introduced new environment variables for Azure Storage configuration: `BlockSize`, `ConcurrentUploadThreads`, and `BlocksBeforeCommit`.
  - Added a new project for testing large file uploads.
  - Implemented a console application for testing file uploads to the Altinn Broker API.

- **Enhancements**
  - Improved error handling and request validation in file transfer processes.
  - Expanded the `ServiceOwnerOverviewExt` class to manage multiple storage providers.
  - Added support for disabling virus scanning during file transfers.
  - Enhanced the `AzureStorageOptions` class for better configuration management.
  - Updated service registrations in dependency injection for improved clarity and functionality.

- **Bug Fixes**
  - Resolved issues with file transfer size validations and error handling for uploads.

- **Documentation**
  - Updated documentation to include instructions for testing large file uploads.

- **Chores**
  - Cleaned up unused code and removed specific file references from project files.

### CODE REVIEW

File: src/Altinn.Broker.Integrations/Azure/AzureResourceManagerService.cs [hunk 1]
```diff
@@ -27,32 +29,48 @@
...
+        if (_hostEnvironment.IsDevelopment())
+        {
+            _backgroundJobClient.Enqueue<IServiceOwnerRepository>(service => service.InitializeStorageProvider(serviceOwnerEntity.Id, "Dummy value", StorageProviderType.Altinn3Azure));
+        } 
+        else
+        {
+            var virusScanStorageProviderJob = _backgroundJobClient.Enqueue<IResourceManager>(service => service.Deploy(serviceOwnerEntity, true, cancellationToken));
+            _backgroundJobClient.ContinueJobWith<IResourceManager>(virusScanStorageProviderJob, service => service.Deploy(serviceOwnerEntity, false, cancellationToken));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply src/Altinn.Broker.Integrations/Azure/AzureResourceManagerService.cs:67]: Exceptions are made as error logs anyway

File: src/Altinn.Broker.Integrations/Azure/AzureResourceManagerService.cs [hunk 2]
```diff
@@ -70,15 +88,17 @@ public async Task Deploy(ServiceOwnerEntity serviceOwnerEntity, CancellationToke
         storageAccountData.Tags.Add("customer_id", serviceOwnerEntity.Id);
         var storageAccountCollection = resourceGroup.Value.GetStorageAccounts();
         var storageAccount = await storageAccountCollection.CreateOrUpdateAsync(WaitUntil.Completed, storageAccountName, storageAccountData, cancellationToken);
-        await EnableMicrosoftDefender(resourceGroupName, storageAccountName, cancellationToken);
+        if (virusScan) { 
+            await EnableMicrosoftDefender(resourceGroupName, storageAccountName, cancellationToken);
+        }
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply src/Altinn.Broker.Integrations/Azure/AzureResourceManagerService.cs:93]: Handled inside EnableMicrosoftDefender

File: src/Altinn.Broker.Integrations/Azure/AzureStorageService.cs [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                $"in {stopwatch.ElapsedMilliseconds / 1000.0:N1}s (avg: {finalSpeedMBps:N2} MB/s)");
+
+            return BitConverter.ToString(blobMd5.Hash).Replace("-", "").ToLowerInvariant();
+        }
+        catch (Exception ex)
+        {
+            logger.LogError("Error occurred while uploading file: {errorMessage}: {stackTrace} ", ex.Message, ex.StackTrace);
+            await blobContainerClient.DeleteIfExistsAsync(cancellationToken: cancellationToken);
+            throw;
+        }
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply src/Altinn.Broker.Integrations/Azure/AzureStorageService.cs]: Good suggestion

File: tests/Altinn.Broker.Tests.LargeFile/Program.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        var httpRequestMessage = new HttpRequestMessage()
+        {
+            RequestUri = new Uri($"https://altinn-testtools-token-generator.azurewebsites.net/api/GetEnterpriseToken?env=tt02&scopes=altinn:broker.write altinn:resourceregistry/resource.write&org=ttd&orgNo={orgNumber}")
+        };
+        var authenticationString = $"{testToolsUsername}:{testToolsPassword}";
+        var base64EncodedAuthenticationString = Convert.ToBase64String(System.Text.Encoding.ASCII.GetBytes(authenticationString));
+        httpRequestMessage.Headers.Authorization = new AuthenticationHeaderValue("Basic", base64EncodedAuthenticationString);
+        var response = await httpClient.SendAsync(httpRequestMessage);
+        var responseContent = await response.Content.ReadAsStringAsync();
+        return responseContent;
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply tests/Altinn.Broker.Tests.LargeFile/Program.cs:82]: Test program does not need to handle errors elegantly

File: tests/Altinn.Broker.Tests.LargeFile/Program.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            Console.WriteLine($"Upload completed in {stopwatch.ElapsedMilliseconds.ToString("N0")} ms");
+        }
+    }
+
+    private static FileTransferInitalizeExt BasicFileTransfer() => new FileTransferInitalizeExt()
+    {
+        ResourceId = testResource,
+        Checksum = null,
+        FileName = "input.txt",
+        PropertyList = [],
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply tests/Altinn.Broker.Tests.LargeFile/Program.cs:147]: Collection expressions were added in C#12

File: src/Altinn.Broker.Core/Domain/ServiceOwnerEntity.cs [hunk 6]
```diff
@@ -1,8 +1,23 @@
...
+﻿using Microsoft.Extensions.Hosting;
+
+namespace Altinn.Broker.Core.Domain;
 
 public class ServiceOwnerEntity
 {
     public required string Id { get; set; }
     public required string Name { get; set; }
+    public List<StorageProviderEntity> StorageProviders { get; set; }
     public StorageProviderEntity? StorageProvider { get; set; }
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply src/Altinn.Broker.Core/Domain/ServiceOwnerEntity.cs]: Fixed

File: src/Altinn.Broker.Integrations/Azure/AzureStorageOptions.cs [hunk 7]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Altinn.Broker.Core.Options;
+
+public class AzureStorageOptions
+{
+    public int BlockSize { get; set; }
+    public int ConcurrentUploadThreads { get; set; }
+    public int BlocksBeforeCommit { get; set; }
+}
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply src/Altinn.Broker.Integrations/Azure/AzureStorageOptions.cs]: Fixed

File: src/Altinn.Broker.API/Models/ServiceOwner/ServiceOwnerOverviewExt.cs [hunk 8]
```diff
@@ -6,5 +6,18 @@ public ServiceOwnerOverviewExt() { }
 
     public required string Name { get; set; }
 
+    public List<StorageProviderExt> StorageProviders { get; set; }
+}
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply src/Altinn.Broker.API/Models/ServiceOwner/ServiceOwnerOverviewExt.cs]: Fixed

File: src/Altinn.Broker.Persistence/Repositories/FileTransferRepository.cs [hunk 9]
```diff
@@ -175,21 +173,21 @@
         }
         var fileTransferId = Guid.NewGuid();
         await using NpgsqlCommand command = dataSource.CreateCommand(
-            "INSERT INTO broker.file_transfer (file_transfer_id_pk, resource_id, filename, checksum, file_transfer_size, external_file_transfer_reference, sender_actor_id_fk, created, storage_provider_id_fk, expiration_time, hangfire_job_id) " +
-            "VALUES (@fileTransferId, @resourceId, @fileName, @checksum, @fileTransferSize, @externalFileTransferReference, @senderActorId, @created, @storageProviderId, @expirationTime, @hangfireJobId)");
-
+            "INSERT INTO broker.file_transfer (file_transfer_id_pk, resource_id, filename, checksum, file_transfer_size, external_file_transfer_reference, sender_actor_id_fk, created, storage_provider_id_fk, expiration_time, hangfire_job_id, use_virus_scan) " +
+            "VALUES (@fileTransferId, @resourceId, @fileName, @checksum, @fileTransferSize, @externalFileTransferReference, @senderActorId, @created, @storageProviderId, @expirationTime, @hangfireJobId, @useVirusScan)");
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply src/Altinn.Broker.Persistence/Repositories/FileTransferRepository.cs:177]: Wrong

File: src/Altinn.Broker.API/Models/FileTransferInitializeExt.cs [hunk 10]
```diff
@@ -64,6 +64,11 @@ public class FileTransferInitalizeExt
     [JsonPropertyName("checksum")]
     [MD5Checksum]
     public string? Checksum { get; set; } = string.Empty;
+
+    /// <summary>
+    /// Disable virus scan. Requires special permission in production.
+    /// </summary>
+    public bool? DisableVirusScan { get; set; } = false;
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply src/Altinn.Broker.API/Models/FileTransferInitializeExt.cs]: Fixed

File: src/Altinn.Broker.Persistence/Repositories/ServiceOwnerRepository.cs [hunk 11]
```diff
@@ -19,20 +19,42 @@
...
                 {
+                    ServiceOwnerId = serviceOwnerId,
                     Created = reader.GetDateTime(reader.GetOrdinal("created")),
                     Id = reader.GetInt64(reader.GetOrdinal("storage_provider_id_pk")),
                     ResourceName = reader.GetString(reader.GetOrdinal("resource_name")),
-                    Type = Enum.Parse<StorageProviderType>(reader.GetString(reader.GetOrdinal("storage_provider_type")))
+                    Type = Enum.Parse<StorageProviderType>(reader.GetString(reader.GetOrdinal("storage_provider_type"))),
+                    Active = reader.GetBoolean(reader.GetOrdinal("active"))
+                };
+
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply src/Altinn.Broker.Persistence/Repositories/ServiceOwnerRepository.cs:47]: Not nullable field tho

File: src/Altinn.Broker.Persistence/Repositories/ServiceOwnerRepository.cs [hunk 12]
```diff
@@ -19,20 +19,42 @@ public class ServiceOwnerRepository(NpgsqlDataSource dataSource) : IServiceOwner
...
         while (reader.Read())
         {
-            serviceOwner = new ServiceOwnerEntity
+            if (serviceOwner == null)
+            {
+                serviceOwner = new ServiceOwnerEntity
+                {
+                    Id = reader.GetString(reader.GetOrdinal("service_owner_id_pk")),
+                    Name = reader.GetString(reader.GetOrdinal("service_owner_name")),
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply src/Altinn.Broker.Persistence/Repositories/ServiceOwnerRepository.cs:31]: Not nullable

File: src/Altinn.Broker.Integrations/DependencyInjection.cs [hunk 13]
```diff
@@ -21,22 +21,20 @@ public static class DependencyInjection
     public static void AddIntegrations(this IServiceCollection services, IConfiguration configuration, bool isDevelopment)
     {
         services.AddSingleton<IResourceManager, AzureResourceManagerService>();
-        services.AddSingleton<IBrokerStorageService, BlobService>();
+        services.AddSingleton<IBrokerStorageService, AzureStorageService>();
```
[the remark this answers is not in the document]
Comment (Ceredron) [reply src/Altinn.Broker.Integrations/DependencyInjection.cs:24]: FIxed

File: src/Altinn.Broker.API/Mappers/FileTransferStatusDetailsExtMapper.cs [hunk 14]
```diff
@@ -11,8 +11,9 @@ internal static FileTransferStatusDetailsExt MapToExternalModel(FileTransferEnti
         var fileTransferOverview = FileTransferStatusOverviewExtMapper.MapToExternalModel(fileTransfer);
         return new FileTransferStatusDetailsExt()
         {
-            Checksum = fileTransferOverview.Checksum,
             FileTransferId = fileTransfer.FileTransferId,
+            ResourceId = fileTransfer.ResourceId,
+            Checksum = fileTransferOverview.Checksum,
             FileName = fileTransferOverview.FileName,
             Sender = fileTransferOverview.Sender,
             FileTransferStatus = fileTransferOverview.FileTransferStatus,
```
Comment (Andreass2) [inline src/Altinn.Broker.API/Mappers/FileTransferStatusDetailsExtMapper.cs:19]: Should we add information about virusscan here, if sender is making the request?

File: src/Altinn.Broker.Application/ApplicationConstants.cs [hunk 15]
```diff
@@ -0,0 +1,9 @@
+namespace Altinn.Broker.Application.Settings;
+
+public static class ApplicationConstants
+{
+    public const long MaxFileUploadSize = 32L * 50000 * 1024 * 1024;
+    public const long MaxVirusScanUploadSize = 2L * 1024 * 1024 * 1024;
+    public const string DefaultGracePeriod = "PT2H";
+    public const string MaxGracePeriod = "PT24H";
+}
```
Comment (Andreass2) [inline src/Altinn.Broker.Application/ApplicationConstants.cs:9]: Nice change!

File: src/Altinn.Broker.Integrations/Azure/AzureStorageService.cs [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+{
+    private async Task<BlobContainerClient> GetBlobContainerClient(FileTransferEntity fileTransferEntity, ServiceOwnerEntity serviceOwnerEntity)
+    {
+        var storageProvider = serviceOwnerEntity.GetStorageProvider(fileTransferEntity.UseVirusScan);
+        var connectionString = await resourceManager.GetStorageConnectionString(storageProvider);
+        var blobServiceClient = new BlobServiceClient(connectionString, new BlobClientOptions()
+        {
+            Retry =
+            {
+                NetworkTimeout = TimeSpan.FromHours(24),
```
Comment (Andreass2) [inline src/Altinn.Broker.Integrations/Azure/AzureStorageService.cs:29]: This means max upload time is 24 hours?

**Review by Andreass2 — APPROVED** [review]
Looks good!

**Comment by Andreass2** [comment]
From docs: 
Would it not make sense for us to update the new maxFileTransferSize at the same time as approval? instead of customer having to do it themselves?

**Comment by Ceredron** [comment]
> From docs: Would it not make sense for us to update the new maxFileTransferSize at the same time as approval? instead of customer having to do it themselves?

Been thinking some more about this, and I think the best would maybe be to never set the max limit on resource at all unless the customer specifically sets it themselves. Originally, we set it to 2GB because that is what the virus scan required. But that limit is better set in code as it is "our" limit, not the resource limit.

File: src/Altinn.Broker.Integrations/Azure/AzureStorageService.cs (hunk 16 above)
Comment (Ceredron) [reply src/Altinn.Broker.Integrations/Azure/AzureStorageService.cs:29]: Yeah, but on further thought, this makes much less sense now actually. This is the network timeout for any single operation. With the block blob client we are using now we are only uploading 32MB at once. If you're spending 24h on that operation you'll probably be killed on minimum rate restriction anyway.

File: src/Altinn.Broker.API/Mappers/FileTransferStatusDetailsExtMapper.cs (hunk 14 above)
Comment (Ceredron) [reply src/Altinn.Broker.API/Mappers/FileTransferStatusDetailsExtMapper.cs:20]: For sure, fixed.

**Pull request merged**